hypothesis-python/src/hypothesis/internal/conjecture/junkdrawer.py

```diff
@@ -413,3 +415,58 @@ def find(self, condition: Callable[[T], bool]) -> T:
                 self.__values.append(value)
                 return value
         raise NotFound("No values satisfying condition")
+
+
+_gc_initialized = False
+_gc_start = 0
+_gc_cumulative_time = 0
+
+
+def _gc_callback(phase, _info):  # pragma: no cover  # called outside coverage by gc
+    global _gc_start, _gc_cumulative_time
+    if phase == "start":
+        _gc_start = time.perf_counter()
+    elif phase == "stop":
+        _gc_cumulative_time += time.perf_counter() - _gc_start
+
+
+def gc_cumulative_time() -> float:
+    global _gc_initialized
+    if not _gc_initialized:
+        if hasattr(gc, "callbacks"):
+            # CPython
+            #
+            # Indirection to simplify monkeypatching of the callback fn
+            # - gc.callbacks can't be reassigned and hence is not suitable for
+            # monkeypatching directly.
+            def gc_callback(phase, info):
+                try:
+                    _gc_callback(phase, info)
+                except RecursionError:  # pragma: no cover
+                    # Avoid flakiness via UnraisableException, which is caught and
+                    # warned by pytest. The actual callback (this function) is
+                    # validated to never trigger a RecursionError itself when
+                    # when called by gc.collect.
```

Comment:
Validation:

```
import gc


def recurse(i=1):
    gc.collect()
    try:
        return recurse(i + 1)
    except RecursionError:
        return i


def inner():
    pass


def gc_callback(phase, info):
    if phase == "start":
        try:
            inner()
        except RecursionError:
            pass


def test_cb_non_raising():
    a = recurse()

    gc.callbacks.append(gc_callback)
    try:
        b = recurse()
    finally:
        gc.callbacks.remove(gc_callback)

    assert a == b
```

Reply:
Oh, sweet - I'll add a test based on this, and I think we'll then be ready to merge 🎉